google/cloud/pubsub/internal/batching_publisher_connection.cc

```diff
@@ -200,9 +198,8 @@ void BatchingPublisherConnection::FlushImpl(std::unique_lock<std::mutex> lk) {
   batch.weak = shared_from_this();
   request.set_topic(topic_full_name_);
 
-  batch_->Flush();
-  batch.batch = batch_;
-  sink_->AsyncPublish(std::move(request)).then(std::move(batch));
+  auto handler = batch_->Flush();
+  sink_->AsyncPublish(std::move(request)).then(std::move(batch)).then(std::move(handler));
```

Comment:
This line makes all your suffering through this long review worth it 😂

Reply:
Yes, it is much cleaner than how it started. I hope it makes your suffering through the review worth it as well 🥳